docs/markdown/snippets/test_dependencies.md

```diff
@@ -0,0 +1,19 @@
+## Test targets no longer built by default
+
+The `ninja test` / `meson test` rules have been reworked to no longer force
+tests to be built unnecessarily. The correct, guaranteed workflow for this has
+been to either run `ninja test` or `meson test`, both of which rebuild
+dependencies on demand. *Also* building test-only binaries as part of
+installing the project (`ninja && ninja install`) was unnecessary and has no
+use case.
+
+Some users might have been relying on the "all" target building test
+dependencies in combination with `meson test --no-rebuild` in order to skip
+calling out to ninja when running tests. The `--no-rebuild` option has always
+been intended for expert usage -- you must provide your own guarantees that it
+will work -- and it should be noted that this change means test programs are no
+longer guaranteed to have been built, depending on whether those test programs
+were *also* defined to build by default / marked as installable. The desired
+behavior of building test programs in a separate stage can be restored by
+building `ninja all meson-test-prereq` (or `meson-benchmark-prereq` for running
+benchmarks), as these prereq targets have been available since meson 0.63.0.
```

Comment:
`ninja test` behavior did not have substantial changes though, did it? Instead the change was in `ninja all` (removing `meson-test-prereq meson-benchmark-prereq`) and `meson test` (removing the invocation of `ninja` with no arguments).

```suggestion
`meson test` and the `ninja all` rule have been reworked to no longer force
unnecessary rebuilds.

`meson test` was invoking `ninja all` due to a bug if the chosen set of tests
had no build dependencies. The behavior is now the same as when tests do
have build dependencies, i.e. to only build the actual set of targets that are
used by the test. This change could cause failures when upgrading to Meson
1.7.0, if the dependencies are not specified correctly in meson.build.

`ninja all` does not rebuild all tests anymore; it should be noted that this
change means test programs are no longer guaranteed to have been
built, depending on whether those test programs were *also* defined
to build by default / marked as installable. This avoids building test-only
binaries as part of installing the project (`ninja && ninja install`), which
is unnecessary and has no use case.  Some users might have been
relying on the "all" target building test dependencies in combination
with `meson test --no-rebuild` in order to skip calling out to ninja when
running tests. The correct, guaranteed workflow for this has always been
to run either `ninja test` or `ninja && meson test`, both of which rebuild
dependencies on demand. 

If you wish to build test programs and dependencies in a separate stage,
you can restore this behavior with `ninja all meson-test-prereq` (or
`meson-benchmark-prereq` for benchmarks), as these prereq targets
have been available since meson 0.63.0.
```

Reply:
Thanks for looking at the wording. I have one critique in particular:

```
Some users might have been
relying on the "all" target building test dependencies in combination
with `meson test --no-rebuild` in order to skip calling out to ninja when
running tests. The correct, guaranteed workflow for this has always been
to run either `ninja test` or `ninja && meson test`, both of which rebuild
dependencies on demand. 
```

In my version, the use of --no-rebuild was defined as "expert usage" in order to lead into the discussion about meson-test-prereq.

Reply:
Yeah, I removed it because my version was already longer and the "expert usage" part didn't really add much. However, the part where there is no guarantee that test dependencies are present is worthwhile. A rephrase could be to break the paragraph before "Some users" and segue into this:

```markdown
Some users might have been relying on the "all" target building test
dependencies in combination with `meson test --no-rebuild` in order to
skip calling out to ninja when running tests. This might break with this
change because, when given `--no-rebuild`, Meson provides no guarantee
that test dependencies are present and up to date. The recommended
workflow is to use either `ninja test` or `ninja && meson test` but, if you
wish to build test programs and dependencies in a separate stage, you can
use for example `ninja all meson-test-prereq meson-benchmark-prereq`
before `meson test --no-rebuild`. These prereq targets have been available
since meson 0.63.0.
```